2671 Move Oral Arguments Search and Alerts to Elasticsearch.

albertisfu opened the pull request:
@mlissner  This PR contains the document index for Oral Arguments. Here are some comments regarding that.

**These are the fields that will be used for search and filtering, according to the current Solr model:**

Available in the search form and advanced search:
`court_id`
`dateArgued` (Date range filter)
`caseName`
`docketNumber`

Only available for advance search:
`id`
`docket_id`
`panel_ids`
`judge`
`dateReargued`
`dateReargumentDenied`


Additionally, I added the following fields that are currently being indexed in Solr, but they are not visible on the [Advanced Search documentation](https://www.courtlistener.com/help/search-operators/#fielded-queries-fieldname-term) for Oral Arguments:

`source`
`duration`
`file_size_mp3`
`download_url`
`local_path`
`court` (docket.court.full_name)

Are these fields for internal use?   or should be included in the advance search documentation?

**I didn't add the following fields, some questions about them:**

`court_citation_string`: I didn't add it since we could get this string from the database during pagination, as in the merge_courts_from_db method of the Parentethical search. 

`court_exact`: I'm not sure if this is needed since it's the same value as court_id. Also, it mentions that it's for faceting, but I'm not sure if we'll have any faceting for Oral Arguments.

`absolute_url`: We won't need this field anymore since we can now use the id and docket_slug to create the audio URL and use the url template tag.

`text`: I saw this field in the Solr model, but I'm not sure what its purpose is and if we'll need it in ES. Is it for Solr to be able to index all the audio fields in a single text field so they can be found in a text search without using advanced search syntax?

I've set this PR to be merged into the Parenthetical Search one in order to avoid the diff from being too large. However, I know that these Oral Arguments PRs will have to wait until Parenthetical Search is merged.

Let me know what you think.

albertisfu commented:
@mlissner This is the big PR (thanks to @ERosendo review and updates) where all the ES code related to Oral Arguments search and alerts.

These are some task that should be completed before deploying this to prod:

- [ ] Create a custom command to index only OA search alerts in the Percolator ES index. (I'll be working on that.)
- [ ] Verify and describe the process to make a smooth transition from OA search and alerts to Elastic. So after we index all the Audio documents and OA alerts, we could just flip the waffle flag or environment variable to make the transition. (I'll be doing some tests to confirm the right process.)
- [ ] I noticed a possible bug related to the random order and ES pagination. I confirmed that using the normal order, either by score or dates, there are no issues with showing the next pages or results. Results can be inconsistent only if new documents are indexed while the user is browsing new pages, so new results can push documents previously shown to a next page. If we want to avoid this possible issue, we could use the API's point-in-time feature as described below.
But using the random order, the order changes with every request, so when requesting the next page, results can be repeated. If that's a problem, we have some options to solve it:
- Introduce a field that includes a random value at indexing time. We can use that field to request a random order. So on each page request, the documents returned will follow a sequential "order," avoiding showing results that were shown on previous pages. However, this alternative won't be completely random if no new documents are added between requests.
- Use a fixed seed for each request when generating the random order. This should be used in combination with the [Point in Time API](https://www.elastic.co/guide/en/elasticsearch/reference/current/point-in-time-api.html), which preserves request parameters within a fixed amount of time. This makes each request performed by a user when browsing pages state-aware, so random results can be held under the duration of that request. This will be more complex to implement since we'll need to generate a point-in-time token on the first page and include it in each subsequent user request.
- [ ] Your review to this PR 

In the meantime, this is not a blocker to continue working on moving RECAP and Opinions to ES. Now it will be easier for @ERosendo and me to work on them since tests are split, and if we need to modify a method, we can do it in this branch and then spread changes easily.

mlissner commented:
I'm not worried about either of those ordering issues. We have the first one already (and nobody complained), and the second one, about randomness, isn't important. It's a rarely used feature, and heck, if page 2 is the same as a refresh of page 1, I think that's fine. We can just document it, so let's do that, and then this is ready for review.

albertisfu commented:
Thanks for your answer, it help me to catch this error. Upon reviewing the API documentation, I realized I had previously overlooked that `random_123` is in fact a way to provide a fixed seed `random_{your_seed}`. So the seed value should be defined by the user.

So considering that the user will provide the random seed, we can use this value to compute the random order and keep it fixed across pagination while the user doesn't change the seed. So the user now won't see repeated values on next pages.

The only difference with the Solr approach is that when Solr parses `random_{seed}` it can accept anything (ints or chars). 
In elasticsearch we can only accept integers since the random order is computed using the formula:
`Long.hashCode(doc['id'].value ^ params.seed)`

So seed needs to be a integer. In case the user provides a not integer seed it uses the current time as seed. That will return random results on each request.

I think we'll need to update [ordering](https://courtlistener.com/help/api/rest/#search-endpoint) documentation to explain that seed should be integer. However, I'm uncertain whether this change should be delayed until all documents have been migrated to Elasticsearch? Or should we do it now?

mlissner commented:
> However, I'm uncertain whether this change should be delayed until all documents have been migrated to Elasticsearch? Or should we do it now?

Making the API documentation more strict now only has upside:

 - Anybody that starts using random ordering now will be ready for ES.
 - We won't forget to do it.

So let's do it. Want to include it here or do a fresh PR?

mlissner commented:
> So let's do it. Want to include it here or do a fresh PR?

Actually, I must insist we do a fresh PR. @david-clark-1043 is updating the API docs now, so this would be quite likely to conflict if we're not careful.

mlissner reviewed (changes requested):
OK, here's my first review. I'll surely do another, but big picture is that this is really great work. 23 comments from me so far, against 7k lines is pretty danged good. A few high level thoughts/questions/reactions...

1. I may have missed this, but can you say a bit about how performance of saving an item will be affected? I'm a little concerned that whenever we save something we're doing too much work. It used to be that:

 - we save an item to the DB
 - we save it to solr in a celery task
 - alerts and webhooks ran later
 
Now it's more like:
 - we save an item to the DB
 - that triggers an update in ES (without a celery task)
 - that triggers an update to the percolator index (without a celery task)
 - that triggers RT alerts and scheduled alert saving
 - RT alert emails get sent (without celery)
  
So I wonder if there are opportunities to make the save faster, and to move the indexing/alerts to one or many tasks?


1. We're doing this a lot:

    ```
    qd = QueryDict(parent_alert.alert.query.encode(), mutable=True)
    search_type = qd.get("type", SEARCH_TYPES.OPINION)
    ```

Let's make a plan to stop that, and indeed we should make a plan to cleanup the alert table completely to make it simpler. Can you file a bug with some notes on that as a post ES cleanup/refactor task?

1. In [your PR about alerts](https://github.com/freelawproject/courtlistener/pull/2823) you said:

> This will enable us to send batched alerts with the original content (even if the document changes in the meantime).

This worries me a bit, actually. What if the item disappears in the meantime, as can happen when we delete docket entries in appellate cases? In that case, if somebody clicks the alert link, it'll give them a 404, right? Maybe this is hard to avoid.

> If a document triggers more than 100 alerts, a new query is performed until there are no more results. This applies to both real-time and batched alerts.

I think we should do this differently. For two reasons. First, if a user has more than 100 results, well, they're using the system wrong. Nobody wants more than 100 results in an alert. Second, I think this opens us up to DOS attacks. Imagine an alert against the query `the` (a very common stopword) that runs monthly on the RECAP Archive. It'd send hundreds of thousands of results. Sad! 

So what I think we should do instead is limit it to 50 (that's still more than anybody should want), and then if there are more than that, put a notice in the alert after the 50th result that says, "Alert emails are limited to 50 results. Please consider refining your query to get more on-point results."

I'm not sure how this will affect webhooks. They only go in realtime, right? If they're scheduled, we'll have to think about that. It's also worth thinking about what happens if somebody sets up a webhook on a stopword query (do we send *every* item?). Hm, I wonder if webhooks need throttles.

> One question, should we add pghistory tracking to these new models UserRateAlert, ParentAlert, and ScheduledAlertHit?

I don't think so. It'd pile up too quickly, I think. We save the sent emails anyway? I'm not sure it'd help too much, and the tables seem ephemeral anyway?

mlissner commented on cl/search/views.py:
```diff
@@ -549,60 +580,69 @@ def es_search(request: HttpRequest) -> HttpResponse:
-    search_form = SearchForm(get_params)
-    document_type = ParentheticalGroupDocument
+    # Add additional or overridden GET parameters
+    if override_params:
```
I forgot this [already bugged me](https://github.com/freelawproject/courtlistener/pull/3026#discussion_r1296647456). At least I'm consistent!

albertisfu replied:
Yeah, I removed `override_params` from this method, instead the dict update is performed before passing to `do_es_search`

albertisfu commented:
Thanks for your comments and suggestions! I'm working on the small changes and refactoring the Alerts models which agree on only using one model for storing scheduled alerts considering we won’t expect having a massive amount of rows in that table so performance now don’t seem an issue.

In addition I wanted to follow up some comments so we can agree on the best approach to follow regarding the following comments.

> I may have missed this, but can you say a bit about how performance of saving an item will be affected? I'm a little concerned that whenever we save something we're doing too much work. It used to be that:

Yeah, I believe performance might actually be an issue, especially in cases where a massive number of new documents are being added (like the recent PACER Santa). If the Elasticsearch cluster can't keep up with all the requests, it will become a problem. So yes, there are some opportunities to move indexing/alerts to tasks.

I followed the current approach thinking about sending real-time (RT) alerts as soon as possible. However, I believe we can make some changes while considering certain trade-offs.

We can index items within a Celery task and still send RT Alerts as soon as the object is stored in the database. Currently, we percolate a document once it's indexed, referring to the document by its index and ID. But we could construct the JSON for the document to be percolated based on the content in the database. This would allow us to percolate it even if it's not yet in Elasticsearch. However, this approach might be risky since alerts could be sent before the document has been indexed. This situation would be quite unusual, but a user might click the alert query link and not see the document in the search results.

However If we can afford some delay in sending RT Alerts, the most performant approach can be something like:

Create a celery task chain that do the following tasks:
`Task: Index document in ES`
`Task: Percolate document after is indexed`
`Task: Send RT Alerts or schedule alerts for other rates.`

In that way we could handle retries and add throttling to tasks.

What do you think about this approach? Can we afford the delay of sending RT alerts through celery tasks?

**About the following comment:**
> qd = QueryDict(parent_alert.alert.query.encode(), mutable=True)
> search_type = qd.get("type", SEARCH_TYPES.OPINION)

> Let's make a plan to stop that, and indeed we should make a plan to cleanup the alert table completely to make it simpler. Can you file a bug with some notes on that as a post ES cleanup/refactor task?

Are you suggesting that we should avoid extracting the search type from the alert query? If that's correct, we could add an `alert_type` field to the Alert table. This field would be set when the user creates the alert. We will need to perform an additional migration step to populate this field for current alerts based on its value in the query. If this is the case, should I apply this change here or later in the issue described below?

> Can you file a bug with some notes on that as a post ES cleanup/refactor task?

Sure, so the issue description would involve reviewing the Alert model to determine if there are any fields that can be removed, correct?

> This worries me a bit, actually. What if the item disappears in the meantime, as can happen when we delete docket entries in appellate cases? In that case, if somebody clicks the alert link, it'll give them a 404, right? Maybe this is hard to avoid.

Got it, yeah, I believe we have some options to address this potential issue.
If we aim to send the latest content of a document in the Alert, we can extract the content from the object in the database and construct the email based on that content.

Additionally, we can verify whether the object already exists in the database before sending alerts for it.

To achieve this, we could introduce a `GenericForeignKey` field to the new single Scheduled Alert model. This would allow us to access the related object, enabling us to build the most recent document to send. This approach would also help us exclude the document from the Alert if it no longer exists.


> I think we should do this differently. For two reasons. First, if a user has more than 100 results, well, they're using the system wrong. Nobody wants more than 100 results in an alert. Second, I think this opens us up to DOS attacks. Imagine an alert against the query the (a very common stopword) that runs monthly on the RECAP Archive. It'd send hundreds of thousands of results. Sad!

yeah I agree, about limiting alert emails to 50 results by alert, so that if an alert is set for a very common word we don’t sent a lot of results in the alert, and we introduce a message to inform users Alerts are limited to 50 results. So I’ll apply this change.

> I'm not sure how this will affect webhooks. They only go in realtime, right? If they're scheduled, we'll have to think about that. It's also worth thinking about what happens if somebody sets up a webhook on a stopword query (do we send every item?). Hm, I wonder if webhooks need throttles.

We also send webhooks for scheduled alerts, so yes, they can also be affected if a user's alert contains a large number of documents. We send one webhook for each triggered alert, not one for each document. Therefore, the webhook will contains a large JSON with all the documents. So we could also implement a limit of 50 documents for each webhook.

Regarding webhooks throttling, for real-time alerts, if we transition to sending alerts as a tasks, we could apply throttling to the task responsible for sending alerts and webhooks. Alternatively, we could create a separate task for sending webhooks and throttle it independently.

For scheduled alerts, we could similarly utilize a task for sending both emails and webhooks and incorporating throttling.


However the code I was describing in the comment:

> If a document triggers more than 100 alerts, a new query is performed until there are no more results. This applies to both real-time and batched alerts.

It's referring to a situation in which a new document is percolated and triggers many alerts for different users. For instance, let’s consider a scenario where 200 users have an alert set up for a very common case. When a new document matching all of these queries is added, the system retrieves the 200 alerts for different users in batches of 100 queries. However, this batch retrieval process can be avoided. We could retrieve all the alerts at once, but we'll need to define a maximum number of alerts that a document could potentially trigger. Do we have an estimate for this maximum number of alerts per document?

mlissner commented:
> > I may have missed this, but can you say a bit about how performance of saving an item will be affected? I'm a little concerned that whenever we save something we're doing too much work. It used to be that:
> 
> Yeah, I believe performance might actually be an issue, especially in cases where a massive number of new documents are being added (like the recent PACER Santa). 

FWIW, that was a lot of *documents*, but we get around 100k new docket entries/day, and that's _way_ more than PACER Santa sent. 

> We can index items within a Celery task and still send RT Alerts as soon as the object is stored in the database. Currently, we percolate a document once it's indexed, referring to the document by its index and ID. But we could construct the JSON for the document to be percolated based on the content in the database. 

Yeah, this is clever, but I think a simpler solution is better, even if it's a bit slower.
 
> However If we can afford some delay in sending RT Alerts, the most performant approach can be something like:
> 
> Create a celery task chain that do the following tasks: `Task: Index document in ES` `Task: Percolate document after is indexed` `Task: Send RT Alerts or schedule alerts for other rates.`

As discussed on Slack, yes, this sounds like a good approach to me.

> **About the following comment:**
> 
> > qd = QueryDict(parent_alert.alert.query.encode(), mutable=True)
> > search_type = qd.get("type", SEARCH_TYPES.OPINION)
> 
> > Let's make a plan to stop that, and indeed we should make a plan to cleanup the alert table completely to make it simpler. Can you file a bug with some notes on that as a post ES cleanup/refactor task?
> 
> Are you suggesting that we should avoid extracting the search type from the alert query? If that's correct, we could add an `alert_type` field to the Alert table. 

Yes, exactly. 

> If this is the case, should I apply this change here or later in the issue described below?

Later. We can do it as clean up after ES is live.

> Sure, so the issue description would involve reviewing the Alert model to determine if there are any fields that can be removed, correct?

Sure. I think there's probably some other cleanup optimizations you'll think of too though, so maybe just do a broad bug for "Post launch Elastic Search code cleanups" or something like that?

> > This worries me a bit, actually. What if the item disappears in the meantime, as can happen when we delete docket entries in appellate cases? In that case, if somebody clicks the alert link, it'll give them a 404, right? Maybe this is hard to avoid.
> 
> Got it, yeah, I believe we have some options to address this potential issue. If we aim to send the latest content of a document in the Alert, we can extract the content from the object in the database and construct the email based on that content.

For now, I think, let's just let this one go. We'll see if it's a real problem and get it fixed down the road if needed.


> I agree about limiting alert emails to 50 results by alert, so that if an alert is set for a very common word we don’t sent a lot of results in the alert, and we introduce a message to inform users Alerts are limited to 50 results. So I’ll apply this change.

Great!

> > I'm not sure how this will affect webhooks. They only go in realtime, right? If they're scheduled, we'll have to think about that. It's also worth thinking about what happens if somebody sets up a webhook on a stopword query (do we send every item?). Hm, I wonder if webhooks need throttles.
> 
> We also send webhooks for scheduled alerts, so yes, they can also be affected if a user's alert contains a large number of documents. We send one webhook for each triggered alert, not one for each document. Therefore, the webhook will contain a large JSON with all the documents. So we could also implement a limit of 50 documents for each webhook.

Thinking about this a bit more, I can't think of a reason why webhooks should ever be hourly/daily/monthly. Servers are always online, and we should just send alerts as they happen. In other words: All webhooks should be real time. Our new infrastructure makes this possible, and we should do it. I'll put a note in the blog about it.

> Regarding webhooks throttling, for real-time alerts, if we transition to sending alerts as a tasks, we could apply throttling to the task responsible for sending alerts and webhooks. Alternatively, we could create a separate task for sending webhooks and throttle it independently.

I think there's a simpler way to do the throttle, and I created an issue for it here: https://github.com/freelawproject/courtlistener/issues/3066
 
> However the code I was describing in the comment:
> 
> > If a document triggers more than 100 alerts, a new query is performed until there are no more results. This applies to both real-time and batched alerts.
> 
> It's referring to a situation in which a new document is percolated and triggers many alerts for different users. For instance, let’s consider a scenario where 200 users have an alert set up for a very common case. When a new document matching all of these queries is added, the system retrieves the 200 alerts for different users in batches of 100 queries. 

Got it. 

> However, this batch retrieval process can be avoided. We could retrieve all the alerts at once, but we'll need to define a maximum number of alerts that a document could potentially trigger. Do we have an estimate for this maximum number of alerts per document?

The batch retrieval is fine and will do the right thing. Thanks for explaining.

albertisfu commented:
Great, thanks for your answers. Just one more question about:

> Thinking about this a bit more, I can't think of a reason why webhooks should ever be hourly/daily/monthly. Servers are always online, and we should just send alerts as they happen. In other words: All webhooks should be real time. Our new infrastructure makes this possible, and we should do it. I'll put a note in the blog about it.

Does that mean that only RT Alerts should trigger webhooks? 
Daily, Weekly and Monthly, shouldn't send webhooks? (so only send emails on scheduled alerts?)

mlissner commented:
> Does that mean that only RT Alerts should trigger webhooks?

No, I was thinking that when an alert is triggered it sends all webhooks immediately, regardless of the alert speed.

albertisfu commented on cl/search/documents.py:
```diff
@@ -89,3 +104,133 @@ def prepare_panel_ids(self, instance):
+        attr="docket.court.full_name",
+        analyzer="text_en_splitting_cl",
+        fields={
+            "exact": fields.TextField(attr="judges", analyzer="english_exact"),
```
Yeah it was wrong, now it's fixed.

albertisfu commented on cl/search/documents.py:
```diff
@@ -89,3 +104,133 @@ def prepare_panel_ids(self, instance):
+        return datetime.utcnow()
+
+
+class PercolatorField(fields.DEDField, Percolator):
```
Yeah, I moved this to `cl.lib.fields` and added a comment:

```
"""Subclass of DEDField and Percolator field.
    This subclass transforms the Percolator field into a field compatible with
    django_elasticsearch_dsl. This enables us to leverage DED Document methods
    like "prepare," which allows the field to be populated using the prepare
    method.
    """
```

   Yes, we'll need this field in other documents.

albertisfu commented on cl/search/documents.py:
```diff
@@ -89,3 +104,133 @@ def prepare_panel_ids(self, instance):
+        qd = QueryDict(instance.query.encode(), mutable=True)
+        search_form = SearchForm(qd)
+        if not search_form.is_valid():
+            return None
```
Yeah, I have added a warning here.

albertisfu commented on cl/search/documents.py:
```diff
@@ -89,3 +104,133 @@ def prepare_panel_ids(self, instance):
+        (
+            query,
+            total_query_results,
+            top_hits_limit,
```
Solved, now these vars are not returned, since I created a new method `build_es_base_query` which doesn't return them. Also this method is used for other basic queries like the Alert estimation query.

albertisfu commented on cl/alerts/management/commands/cl_send_scheduled_alerts.py:
```diff
@@ -0,0 +1,123 @@
+# !/usr/bin/python
+# -*- coding: utf-8 -*-
```
Removed.

albertisfu commented on cl/alerts/models.py:
```diff
@@ -141,3 +144,64 @@ class RealTimeQueue(models.Model):
     item_pk = models.IntegerField(help_text="the pk of the item")
+
+
+class UserRateAlert(AbstractDateTimeModel):
```
These models are now removed. Only ScheduledAlertHit remains and added docstrings.

albertisfu commented on cl/search/views.py:
```diff
@@ -456,6 +475,19 @@ def show_results(request: HttpRequest) -> HttpResponse:
                 if request.GET.get("type") == SEARCH_TYPES.PARENTHETICAL:
                     search_results = do_es_search(request.GET.copy())
                     render_dict.update(search_results)
+                elif request.GET.get("type") == SEARCH_TYPES.ORAL_ARGUMENT:
```
Yeah, I have applied this suggestion and removed the `elif`

albertisfu commented on cl/search/views.py:
```diff
@@ -549,60 +580,69 @@ def es_search(request: HttpRequest) -> HttpResponse:
-    :return: dict
+    :param rows: The number of solr results to request
+    :param override_params: A dict with additional or different GET params to
+    be sent to solr.
```
Fixed.

albertisfu commented on cl/search/views.py:
```diff
@@ -632,6 +686,14 @@ def do_es_pagination(get_params, hits, rows_per_page=5):
+    if error:
+        return [], query_time, error
+    if get_params.get("type") == SEARCH_TYPES.PARENTHETICAL:
+        hits = hits.aggregations.groups.buckets
```
Yeah, I moved this to `fetch_es_results` and now is handled automatically:

```
        if response.aggregations:
            response = response.aggregations.groups.buckets
```

albertisfu commented on cl/search/signals.py:
```diff
@@ -1,5 +1,10 @@
+from django.dispatch import receiver
+
+from cl.alerts.send_alerts import send_or_schedule_alerts
+from cl.audio.models import Audio
```
These are not required anymore, removed.

albertisfu commented on cl/search/elasticsearch_files/synonyms_en.txt:
```diff
@@ -0,0 +1,762 @@
+zam,zambia
+zim,zimbabwe
+
+# Top abbreviations in legal documents since 1975
```
Done, added to `synonyms_en.txt`

albertisfu commented on cl/audio/models.py:
```diff
@@ -26,6 +27,7 @@
 class Audio(AbstractDateTimeModel):
     """A class representing oral arguments and their associated metadata"""
 
+    SUPPORT_ALERTS = True
```
Yeah, I agree, I've removed this one and replaced for a list in `cl.search.tasks`

`models_alert_support = [Audio] `

So we'll add more Models here for other documents.

albertisfu commented on cl/lib/elasticsearch_utils.py:
```diff
@@ -133,32 +284,50 @@ def build_es_filters(cd: CleanData) -> List:
+            build_text_filter("caseName", cd.get("case_name", ""))
+        )
+        # Build court terms filter
+        queries_list.extend(build_text_filter("judge", cd.get("judge", "")))
```
Fixed.

albertisfu commented on cl/api/views.py:
```diff
@@ -171,32 +175,49 @@ async def get_result_count(request, version, day_count):
+        cd["argued_before"] = None
+
+        search_query = document_type.search()
+        s, total_query_results, top_hits_limit = build_es_main_query(
```
Yeah, now this is using a more simpler method: `build_es_base_query`

albertisfu commented on cl/lib/es_signal_processor.py:
```diff
@@ -94,11 +106,19 @@ def save_document_in_es(
         refresh=settings.ELASTICSEARCH_DSL_AUTO_REFRESH,
     )
+    support_alerts = getattr(instance, "SUPPORT_ALERTS", None)
+    if support_alerts and response["_version"] == 1:
```
Sure, I added the comment, now this method is a task in `cl.search.tasks`

albertisfu commented on cl/alerts/send_alerts.py:
```diff
@@ -0,0 +1,265 @@
+    :param document_id: The document ID in ES index to be percolated.
+    :param document_index: The ES document index where the document lives.
+    :param search_after: The ES search_after param for deep pagination.
+    :return: The response from the Elasticsearch query or an empty list if an
```
Yeah, I refactored this method `percolate_document` , now it's in `cl.alerts.utils`

albertisfu commented on cl/alerts/send_alerts.py:
```diff
@@ -0,0 +1,265 @@
+
+        # Schedule no RT Alerts.
+        scheduled_rates = [Alert.DAILY, Alert.WEEKLY, Alert.MONTHLY]
+        if alert_triggered.rate in scheduled_rates:
```
Yeah, I applied the suggestion, also this method `process_percolator_response` was refactored as a task with some improvements in order to increase the performance. Now it lives in `cl.alerts.tasks`

albertisfu commented on cl/alerts/send_alerts.py:
```diff
@@ -0,0 +1,265 @@
+                except IntegrityError:
+                    user_rate = UserRateAlert.objects.get(
+                        user=alert_triggered.user, rate=alert_triggered.rate
+                    )
```
Well, this was removed, now we only create one instance for each hit triggered and stored in the model `ScheduledAlertHit`.

albertisfu commented on cl/alerts/management/commands/cl_send_scheduled_alerts.py:
```diff
@@ -0,0 +1,123 @@
+
+    # Remove stored alerts sent, all the related objects will be deleted
+    # in cascade.
+    rate_users.delete()
```
This command was also refactored. I realized that it's better to only change the status of the `ScheduledAlertHit` instances to `Sent` and only removed them after some time (I set it to 90 days). So in case of debugging we'll have the object in DB.

albertisfu commented on cl/alerts/management/commands/cl_send_scheduled_alerts.py:
```diff
@@ -0,0 +1,123 @@
+                        result.highlighted_fields,
+                        document_content,
+                        ALERTS_HL_TAG,
+                    )
```
Yeah, now the document is stored with highlights included.

albertisfu commented on cl/alerts/models.py:
```diff
@@ -141,3 +144,64 @@ class RealTimeQueue(models.Model):
     item_pk = models.IntegerField(help_text="the pk of the item")
+
+
+class UserRateAlert(AbstractDateTimeModel):
```
I've applied these refactor, now we only use one model `ScheduledAlertHit` to store scheduled hits. When sending them using the command `cl_send_scheduled_alerts` hits are grouped by user and alert instead.

albertisfu commented:
@mlissner This is ready for another review,  I've applied them, some comments above.

Additionally in order to increase performance when saving documents in ES and send alerts:
- I've moved all the Elasticsearch indexing/updating document methods to Celery.
- Also the percolator query was also moved to celery.
- Search alert webhooks and alert emails also moved to celery.

The Alert model now has a field: `alert_type` which is populated on `save()` according to the stored query.
So we can avoid doing this:

```
qd = QueryDict(parent_alert.alert.query.encode(), mutable=True)
search_type = qd.get("type", SEARCH_TYPES.OPINION)

```
Every time we need the `search_type`.

- Now all search Alert Webhooks are sent in real time, regardless of the Alert rate or user donations.
- Scheduled Alerts only trigger emails.


I performed some manual tests just to confirm everything was working as expected now that ES methods are celery tasks and also confirm the UI and alert templates looks fine and I found some issues that are now solved:

-  ES `date` fields can contain `date` or `datetime` so the `SearchESResultSerializer` was failing when receiving a `date` since the field was mapped to `datetime`. So I introduced a custom field `DateOrDateTimeField` so it an support date or datime.
- Disabled Alerts `OFF` were triggering and scheduled Alerts. So I fixed excluding `OFF` rates from the percolator query.
- Also fixed the Edit Alert form that was not being rendered.
- The View original Court URL is not shown if `download_url` is None in the HTML template.
- `ScheduledAlertHit` SENT instances are not removed immediately but 90 days after.

Thanks for your comments and suggestions.

Let me know what do you think.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -406,3 +429,281 @@ def send_recap_email_user_not_found(recap_email_recipients: list[str]) -> None:
+
+    alerts_triggered.extend(percolator_response.hits)
+
+    # Check if the query contains more documents than PERCOLATOR_PAGE_SIZE.
+    # If so, return additional results until there are not more.
```
```suggestion
    # Check if the query contains more documents than PERCOLATOR_PAGE_SIZE.
    # If so, return additional results until there are not more.
    # Remember, percolator results are alerts, not documents, so what you're 
    # paginating are user alerts that the document matched, not documents that
    # an alert matched. 🙃.
```

albertisfu replied:
changed.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -406,3 +429,281 @@ def send_recap_email_user_not_found(recap_email_recipients: list[str]) -> None:
+
+    scheduled_hits_to_create = []
+    email_alerts_to_send = []
+    rt_alerts_send = []
```
```suggestion
    rt_alerts_to_send = []
```

albertisfu replied:
changed.

mlissner reviewed (changes requested):
This looks great. Nine small things for you, and I think we can merge this on Thursday. Big day.

mlissner commented:
And, sorry, one other thing. You've got a save() method for populating the new field on the alerts, BUT how does it get populated for existing alerts? In a future PR, let's also make rate a required field, and tweak our code to have it as part of wherever we do `Alert.object.create()`.

mlissner commented:
And one more thing. I was at dinner last night, and somebody casually mentioned `edpa`, and I realized synonyms are important enough to just do it for all the four-letter court abbreviations. 

Therefore, can you please add all of these as well:

```
mdal,almd
ndal,alnd
sdal,alsd
edar,ared
wdar,arwd
cdca,cacd
edca,caed
ndca,cand
sdca,casd
mdfl,flmd
ndfl,flnd
sdfl,flsd
mdga,gamd
ndga,gand
sdga,gasd
cdil,ilcd
ndil,ilnd
sdil,ilsd
ndin,innd
sdin,insd
ndia,iand
sdia,iasd
edky,kyed
wdky,kywd
edla,laed
mdla,lamd
wdla,lawd
edmi,mied
wdmi,miwd
ndms,msnd
sdms,mssd
edmo,moed
wdmo,mowd
edny,nyed
ndny,nynd
sdny,nysd
wdny,nywd
ednc,nced
mdnc,ncmd
wdnc,ncwd
ndoh,ohnd
sdoh,ohsd
edok,oked
ndok,oknd
wdok,okwd
edpa,paed
mdpa,pamd
wdpa,pawd
edtn,tned
mdtn,tnmd
wdtn,tnwd
edtx,txed
ndtx,txnd
sdtx,txsd
wdtx,txwd
edva,vaed
wdva,vawd
edwa,waed
wdwa,wawd
ndwv,wvnd
sdwv,wvsd
edwi,wied
wdwi,wiwd
```

albertisfu commented on cl/alerts/models.py:
```diff
@@ -146,62 +158,52 @@ class RealTimeQueue(models.Model):
+    rate = models.CharField(
+        help_text="The rate chosen by the user for the alert",
+        choices=Alert.FREQUENCY,
+        max_length=10,
```
yeah, I forgot to remove this field from here. Now removed and the rate is retrieved from `alert`.

albertisfu commented on cl/alerts/management/commands/cl_send_scheduled_alerts.py:
```diff
@@ -38,57 +37,52 @@ def query_and_send_alerts_by_rate(rate: str) -> None:
-        send_search_alert_and_webhooks(rate_user.user, hits)
-        if rate_user.parent_alerts.exists():
+        send_search_alert_emails.delay([(user_id, hits)])
+        if hits:
```
Yeah, this makes sense, change it!

albertisfu commented on cl/alerts/management/commands/cl_send_scheduled_alerts.py:
```diff
@@ -38,57 +37,52 @@ def query_and_send_alerts_by_rate(rate: str) -> None:
-    # in cascade.
-    rate_users.delete()
+    # Remove old Scheduled alert hits sent.
+    delete_old_scheduled_alerts()
```
Sure, this will be only executed on daily rate.

albertisfu commented on cl/alerts/management/commands/cl_send_scheduled_alerts.py:
```diff
@@ -107,6 +101,19 @@ def send_scheduled_alerts(rate: str) -> None:
+    scheduled_hits_to_delete = ScheduledAlertHit.objects.filter(
+        date_created__lt=older_than, hit_status=SCHEDULED_ALERT_HIT_STATUS.SENT
+    ).delete()
+    return scheduled_hits_to_delete[0]
```
Sure, I have tweaked the method, so SENT `ScheduledAlertHit` are moved after `DAYS_TO_DELETE` and SCHEDULED `ScheduledAlertHit` are removed after `2 * DAYS_TO_DELETE`.

albertisfu commented on cl/alerts/migrations/0008_add_scheduled_alert_hit_and_alert_type_to_alert_customer.sql:
Yeah, I'm following the suggestions we [discussed](https://github.com/freelawproject/courtlistener/pull/2477/files#r1109949256) some time ago: 

> 0013_add_docket_entry_time_filed.py with everything.
> 0013_add_docket_entry_time_filed.sql the sql migration for our replica, only with the regular and event tables.
> 0013_add_docket_entry_time_filed_customers.sql the sql migration for customers, only with the regular table.

So the difference in customers file is that it doesn't have event tables migrations.

mlissner replied:
Love it, thank you. Can you update the wiki with this too, please, when you get a moment?

albertisfu commented on cl/search/api_serializers.py:
```diff
@@ -306,11 +321,20 @@ def create(self, validated_data):
         "text": serializers.CharField,
         "keyword": serializers.CharField,
-        "date": serializers.DateTimeField,
+        "date": DateOrDateTimeField,
```
yeah, this is not required anymore, as we talked about, for now we'll go with the current Datetime format for date fields. When working on the new Search API endpoints we'll consider using naive datetimes for values that doesn't have a timezone.

I didn't require to use `midnight_pt` since by default `serializers.DateTimeField` adds the system timezone in this case (PT).

So date fields preserver the same format:

```
 {
            "absolute_url": "/audio/10/nelson-plc-v-gomez-bell-and-moreno/",
            "caseName": "Nelson PLC v. Gomez, Bell and Moreno",
            "court": "Thirteenth circuit of the Eruptanyom",
            "court_citation_string": "",
            "court_exact": "tuubd",
            "court_id": "tuubd",
            "dateArgued": "1971-02-07T00:00:00-08:00",
            "dateReargued": null,
            "dateReargumentDenied": null,
            "docketNumber": "4:05-ms-35426",
            "docket_id": 10,
            "download_url": "http://rose.net/",
            "duration": null,
            "file_size_mp3": 0,
            "id": 10,
            "judge": null,
            "local_path": "dat/1971/02/07/example_4.dat",
            "pacer_case_id": "313861",
            "panel_ids": [],
            "sha1": "4d6649a0ee4312a86c12147d0a275b59751f892f",
            "snippet": null,
            "source": "M",
            "timestamp": "2023-08-31T19:23:29.460227-07:00"
        }
```

albertisfu replied:
Ok, this changed again. As we talked about, this also impact on webhooks. 

On webhooks we'll start showing date values as dates instead of transforming them to datetime fields (e.g: dateArgued, dateFiled, etc).

On the API, current version will continue showing date fields with datetime values using the system timezone (PT).

datetime fields will show datetime values e.g: timestamp.

So I added an improved version of `DateOrDateTimeField`

So according to the field value received it ill render a `date` or `datetime`.

The difference we have between the API and webhooks is that when we retrieve results from Elasticsearch, we get a naive `datetime` value for all date fields. However, for sending webhooks, we don’t perform an additional query to the OA index in ES to fetch the document. Instead, we save the document in the database, prepare it (extracting the fields we usually send to ES when indexing the document), and then index it. If an alert is triggered, we use this prepared version to send alerts. This version contains fields as they are in the database. As a result, date fields are in `date` format, not in `datetime` format as they are when retrieved from ES.

albertisfu commented:
Thanks!  requested changes are ready! some comments above.

The synonyms file is now updated with all the courts synonyms.

> And, sorry, one other thing. You've got a save() method for populating the new field on the alerts, BUT how does it get populated for existing alerts? In a future PR, let's also make rate a required field, and tweak our code to have it as part of wherever we do Alert.object.create().

Yeah, I created https://github.com/freelawproject/courtlistener/issues/3080 describing this tweak and other after ES launch. 

For now in order to populate the `alert_type` field on existing Alerts we can do it running the following script before start indexing Alert into ES:

```
from cl.alerts.models import Alert
from django.db import transaction

all_alerts = Alert.objects.all()
for alert in all_alerts:
    alert.save()
```

mlissner reviewed (approved):
Approved. 

 - 250 commits
 - 92 comments
 - 8,429 new/changed lines, 747 removed

Thank you everybody.  We'll merge and deploy tomorrow.

mlissner commented:
This is now deployed, with the content slowly coming in. If you are a super user, you will see the new oral arguments search interface! Everybody is now getting oral argument alerts from Elastic (the first feature exposed to the public).

Hooray!
